ext/dom/xpath_rust.rs

```diff
@@ -0,0 +1,56 @@
+use std::ffi::{CString, c_char};
+
+
+#[no_mangle]
+pub extern "C" fn domxpath_quote_literal(input: *const c_char, len: *mut usize) -> *mut c_char {
+    let slice = unsafe { std::slice::from_raw_parts(input as *const u8, *len as usize) };
+
+    let single_quote_absent = !slice.contains(&b'\'');
+    let double_quote_absent = !slice.contains(&b'"');
+
+    let result = if single_quote_absent {
+        let mut res = Vec::with_capacity(slice.len() + 2);
```

Comment:
Without additional infrastructure this bypasses PHP's memory manager and thus PHP's memory limit.

Reply:
Good point. I wonder how difficult it is to give rust access to php's memory manager 🤔

Reply:
Using GlobalAlloc makes sense, request-based allocations are the most common case anyway.
If you go for an RFC then we'd need more than just the allocation interface, but we also need proper access to the Zend data structures. So this is more work than we might expect at first, but my personal opinion is that it's a good idea to allow memory safe languages inside PHP.

Reply:
@danog please do. I won't have time/energy to contribute for some time.